File: api/routes/backtest.py
"""POST /api/backtest — бэктест стратегии."""
import os, sys, asyncio
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from litestar import post
from litestar.exceptions import HTTPException

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))

from api.routes.candles    import get_client, _executor
from api.routes.validation import (
    validate_ticker, validate_formula, validate_days,
    validate_capital, validate_name,validate_interval, validate_size,
)
from strategy.base     import FormulaStrategy
from strategy.backtest import run_backtest

MAX_DAYS = {
    "1m": 1, "5m": 3, "15m": 7, "1h": 30, "1d": 365,
}


# ── Датаклассы ────────────────────────────────────────────

@dataclass
class BacktestRequest:
    ticker:        str
    name:          str
    entry_formula: str
    exit_formula:  str
    stop_formula:  Optional[str]  = None
    size:          float          = 10.0
    interval:      str            = "1d"
    days:          Optional[int]  = None
    capital:       float          = 100_000.0
    params:        Dict[str, Any] = field(default_factory=dict)


@dataclass
class TradePoint:
    entry_time:  int
    exit_time:   int
    entry_price: float
    exit_price:  float
    action:      str
    pnl:         float
    pnl_pct:     float
    size:        float


@dataclass
class EquityPoint:
    time:   int
    equity: float


@dataclass
class BacktestResponse:
    ticker:       str
    name:         str
    trades:       List[TradePoint]
    equity_curve: List[EquityPoint]
    initial:      float
    final:        float
    total_return: float
    max_drawdown: float
    sharpe:       float
    winrate:      float
    total_trades: int
    wins:         int
    losses:       int
    avg_profit:   float
    avg_loss:     float
    best_trade:   float
    worst_trade:  float
    commission:   float
    error:        Optional[str] = None


# ── Логика ────────────────────────────────────────────────

def _run_bt(req: BacktestRequest):
    ticker  = validate_ticker(req.ticker)
    entry   = validate_formula(req.entry_formula, 'entry_formula')
    exit_   = validate_formula(req.exit_formula,  'exit_formula')
    stop    = validate_formula(req.stop_formula,  'stop_formula')
    days = validate_days(req.days) or MAX_DAYS.get(req.interval, 365)
    capital = validate_capital(req.capital)
    name    = validate_name(req.name)
    interval = validate_interval(req.interval)
    size     = validate_size(req.size) 
    client   = get_client()
    figi     = client.find_figi(ticker)
    df       = client.get_candles(figi=figi, interval=req.interval, days_back=days)

    if df is None or df.empty:          # ← добавить
        raise ValueError(f"Нет данных для {ticker} / {req.interval}")

    strategy = FormulaStrategy(
        name          = name,
        entry_formula = entry,
        exit_formula  = exit_,
        stop_formula  = stop,
        size          = req.size,
        interval      = req.interval,
        params        = req.params,
    )
    return run_backtest(strategy, df, initial_capital=capital)


# ── Endpoint ──────────────────────────────────────────────

@post("/backtest")
async def run_backtest_endpoint(data: BacktestRequest) -> BacktestResponse:
    try:
        loop   = asyncio.get_event_loop()
        result = await loop.run_in_executor(_executor, lambda: _run_bt(data))
    except (SyntaxError, ValueError, RuntimeError) as e:
        return BacktestResponse(
            ticker=data.ticker, name=data.name,
            trades=[], equity_curve=[],
            initial=data.capital, final=data.capital,
            total_return=0, max_drawdown=0, sharpe=0,
            winrate=0, total_trades=0, wins=0, losses=0,
            avg_profit=0, avg_loss=0, best_trade=0,
            worst_trade=0, commission=0, error=str(e)
        )
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

    return BacktestResponse(
        ticker       = data.ticker.upper(),
        name         = data.name,
        trades       = [TradePoint(
            entry_time  = t.entry_time,  exit_time   = t.exit_time,
            entry_price = t.entry_price, exit_price  = t.exit_price,
            action      = t.action,      pnl         = t.pnl,
            pnl_pct     = t.pnl_pct,     size        = t.size,
        ) for t in result.trades],
        equity_curve = [
            EquityPoint(time=e.time, equity=e.equity)
            for e in result.equity_curve
        ],
        initial      = result.initial,
        final        = result.final,
        total_return = result.total_return,
        max_drawdown = result.max_drawdown,
        sharpe       = result.sharpe,
        winrate      = result.winrate,
        total_trades = result.total_trades,
        wins         = result.wins,
        losses       = result.losses,
        avg_profit   = result.avg_profit,
        avg_loss     = result.avg_loss,
        best_trade   = result.best_trade,
        worst_trade  = result.worst_trade,
        commission   = result.commission,
        error        = None,
    )

File: frontend/backtest.js
// ── Состояние бэктеста ────────────────────────────────────
const btState = {
    result:     null,
    equityChart: null,
};

// ── Шаблоны (те же что в стратегиях) ─────────────────────
const BT_TEMPLATES = {
    ema_cross:    { name: 'EMA Cross',    entry: 'RESULT = CROSS_UP(EMA(9), EMA(21))',    exit: 'RESULT = CROSS_DOWN(EMA(9), EMA(21))',  stop: '' },
    rsi_reversal: { name: 'RSI Reversal', entry: 'RESULT = CROSS_UP(RSI(14), 30)',        exit: 'RESULT = CROSS_DOWN(RSI(14), 70)',      stop: 'RESULT = RSI(14) < 20' },
    trend_follow: { name: 'Trend Follow', entry: 'RESULT = EMA(9) > EMA(21) AND CROSS_UP(CLOSE, EMA(9))', exit: 'RESULT = CROSS_DOWN(CLOSE, EMA(21))', stop: '' },
    bb_bounce:    { name: 'BB Bounce',    entry: 'RESULT = CLOSE < BB_LOWER(20, 2)',      exit: 'RESULT = CLOSE > BB_UPPER(20, 2)',      stop: '' },
};

function setBtTemplate(key) {
    const t = BT_TEMPLATES[key];
    document.getElementById('bt-name').value  = t.name;
    document.getElementById('bt-entry').value = t.entry;
    document.getElementById('bt-exit').value  = t.exit;
    document.getElementById('bt-stop').value  = t.stop;
}

function initBacktest() {
    const el = document.getElementById('bt-templates');
    if (!el) return;
    el.innerHTML = Object.entries(BT_TEMPLATES).map(([key, t]) =>
        `<button class="btn btn-tpl" onclick="setBtTemplate('${key}')">${t.name}</button>`
    ).join('');
}

// ── Запуск бэктеста ───────────────────────────────────────
async function runBacktest() {
    const ticker   = document.getElementById('ticker').value.trim().toUpperCase();
    const name     = document.getElementById('bt-name').value.trim()   || 'Backtest';
    const entry    = document.getElementById('bt-entry').value.trim();
    const exit_    = document.getElementById('bt-exit').value.trim();
    const stop     = document.getElementById('bt-stop').value.trim();
    const capital  = parseFloat(document.getElementById('bt-capital').value)  || 100000;
    const interval = document.getElementById('bt-interval').value;
    const days     = parseInt(document.getElementById('bt-days').value) || 365;

    if (!entry) { setBtStatus('❌ Введите Entry формулу'); return; }
    if (!exit_) { setBtStatus('❌ Введите Exit формулу');  return; }

    setBtStatus(`⏳ Тестируем ${ticker} за ${days} дней...`);
    document.getElementById('btn-run-bt').disabled = true;

    try {
        const r = await fetch('/api/backtest', {
            method:  'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                ticker,
                name,
                entry_formula: entry,
                exit_formula:  exit_,
                stop_formula:  stop || null,
                size:          parseFloat(document.getElementById('bt-size').value) || 10,
                interval,
                days,
                capital,
                params: {},
            })
        });

        const data = await r.json();

        if (!r.ok) {
            console.error('[Backtest] Server error:', data);   // ← увидишь detail
            setBtStatus(`❌ ${data.detail || 'Ошибка сервера ' + r.status}`);
            return;
        }

        if (data.error) { setBtStatus(`❌ ${data.error}`); return; }

        btState.result = data;

        renderBtStats(data);
        renderBtTrades(data.trades);
        renderEquityChart(data.equity_curve, data.initial);

        const sign = data.total_return >= 0 ? '+' : '';
        setBtStatus(
            `✅ ${data.total_trades} сделок | ` +
            `${sign}${data.total_return}% | ` +
            `Winrate: ${data.winrate}% | ` +
            `Sharpe: ${data.sharpe} | ` +
            `Просадка: -${data.max_drawdown}%`
        );

    } catch (e) {
        setBtStatus(`❌ ${e.message}`);
    } finally {
        document.getElementById('btn-run-bt').disabled = false;
    }
}

// ── Метрики ───────────────────────────────────────────────
function renderBtStats(d) {
    const el = document.getElementById('bt-stats');
    if (!el) return;

    const returnColor = d.total_return >= 0 ? 'var(--green)' : 'var(--red)';
    const sign        = d.total_return >= 0 ? '+' : '';

    el.innerHTML = `
        <div class="bt-metrics">
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:${returnColor}">${sign}${d.total_return}%</div>
                <div class="bt-metric-lbl">Доходность</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val">${d.final.toLocaleString('ru')} ₽</div>
                <div class="bt-metric-lbl">Итоговый капитал</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:var(--green)">${d.winrate}%</div>
                <div class="bt-metric-lbl">Winrate</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val">${d.total_trades}</div>
                <div class="bt-metric-lbl">Сделок</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:var(--red)">-${d.max_drawdown}%</div>
                <div class="bt-metric-lbl">Макс просадка</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val">${d.sharpe}</div>
                <div class="bt-metric-lbl">Sharpe</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:var(--green)">+${d.avg_profit}%</div>
                <div class="bt-metric-lbl">Avg прибыль</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:var(--red)">${d.avg_loss}%</div>
                <div class="bt-metric-lbl">Avg убыток</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:var(--green)">+${d.best_trade}%</div>
                <div class="bt-metric-lbl">Лучшая сделка</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val" style="color:var(--red)">${d.worst_trade}%</div>
                <div class="bt-metric-lbl">Худшая сделка</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val">${d.wins} / ${d.losses}</div>
                <div class="bt-metric-lbl">Прибыльных / убыточных</div>
            </div>
            <div class="bt-metric">
                <div class="bt-metric-val">${d.commission.toLocaleString('ru')} ₽</div>
                <div class="bt-metric-lbl">Комиссия (0.05%)</div>
            </div>
        </div>
    `;
}

// ── Таблица сделок ────────────────────────────────────────
function renderBtTrades(trades) {
    const el = document.getElementById('bt-trades');
    if (!el) return;

    if (!trades.length) {
        el.innerHTML = '<div class="muted" style="padding:12px">Сделок нет</div>';
        return;
    }

    const header = `
        <div class="bt-trade-row bt-trade-header">
            <span>Вход</span>
            <span>Выход</span>
            <span>Цена входа</span>
            <span>Цена выхода</span>
            <span>P&L ₽</span>
            <span>P&L %</span>
            <span>Тип</span>
        </div>
    `;

    const rows = [...trades].reverse().map(t => {
        const entryDt  = new Date(t.entry_time * 1000).toLocaleString('ru');
        const exitDt   = new Date(t.exit_time  * 1000).toLocaleString('ru');
        const pnlColor = t.pnl >= 0 ? 'var(--green)' : 'var(--red)';
        const sign     = t.pnl >= 0 ? '+' : '';
        const icon     = t.action === 'STOP' ? '🛑' : t.pnl >= 0 ? '✅' : '❌';

        return `
            <div class="bt-trade-row">
                <span>${entryDt}</span>
                <span>${exitDt}</span>
                <span>${t.entry_price} ₽</span>
                <span>${t.exit_price} ₽</span>
                <span style="color:${pnlColor}">${sign}${t.pnl} ₽</span>
                <span style="color:${pnlColor}">${sign}${t.pnl_pct}%</span>
                <span>${icon} ${t.action}</span>
            </div>
        `;
    }).join('');

    el.innerHTML = header + rows;
}

// ── График кривой капитала (LightweightCharts) ────────────
function renderEquityChart(equityCurve, initial) {
    const container = document.getElementById('bt-equity-chart');
    if (!container || !equityCurve.length) return;

    container.innerHTML = '';

    if (btState.equityChart) {
        try { btState.equityChart.remove(); } catch(_) {}
        btState.equityChart = null;
    }

    const chart = LightweightCharts.createChart(container, {
        width:  container.clientWidth,
        height: 220,
        layout: { background: { color: '#131722' }, textColor: '#d1d4dc' },
        grid:   { vertLines: { color: '#1e222d' }, horzLines: { color: '#1e222d' } },
        timeScale:       { borderColor: '#363a45', timeVisible: true },
        rightPriceScale: { borderColor: '#363a45' },
        crosshair: { mode: 1 },
    });

    btState.equityChart = chart;

    // Линия капитала
    const line = chart.addLineSeries({
        color:     '#2196F3',
        lineWidth: 2,
        title:     'Капитал',
    });

    // Базовая линия (initial)
    const baseline = chart.addLineSeries({
        color:     '#363a45',
        lineWidth: 1,
        lineStyle: 2,   // dashed
        title:     'Начало',
    });

    const points = equityCurve.map(e => ({ time: e.time, value: e.equity }));
    points.sort((a, b) => a.time - b.time);

    line.setData(points);
    baseline.setData([
        { time: points[0].time,  value: initial },
        { time: points.at(-1).time, value: initial },
    ]);

    chart.timeScale().fitContent();

    // Resize
    new ResizeObserver(() => {
        chart.applyOptions({ width: container.clientWidth });
    }).observe(container);
}

function setBtStatus(msg) {
    const el = document.getElementById('bt-status');
    if (el) el.textContent = msg;
}
